File: apps/auth-server/src/main/java/com/example/chat/auth/server/api/AuthApi.java
package com.example.chat.auth.server.api;

import com.example.chat.auth.server.api.dto.factory.CredentialFactory;
import com.example.chat.auth.server.api.dto.request.AuthenticateRequest;
import com.example.chat.auth.server.api.dto.request.CompleteMfaRequest;
import com.example.chat.auth.server.api.dto.request.SignupRequest;
import com.example.chat.auth.server.api.dto.response.AuthResponse;
import com.example.chat.auth.server.api.dto.response.SignupResponse;
import com.example.chat.auth.server.application.service.AuthenticationApplicationService;
import com.example.chat.auth.server.application.service.MfaApplicationService;
import com.example.chat.auth.server.application.service.SignupApplicationService;
import com.example.chat.auth.server.core.domain.AuthenticationContext;
import com.example.chat.auth.server.core.domain.Credential;
import com.example.chat.auth.server.core.domain.CredentialType;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseCookie;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.Instant;

@Slf4j
@RestController
@RequestMapping("/api/v1/auth")
@RequiredArgsConstructor
public class AuthApi {

    private final AuthenticationApplicationService authenticationService;
    private final MfaApplicationService mfaService;
    private final CredentialFactory credentialFactory;
    private final SignupApplicationService signupService;

    private static final String REFRESH_TOKEN_COOKIE = "refresh_token";

    /** 회원가입 */
    @PostMapping("/signup")
    public ResponseEntity<SignupResponse> signup(@Valid @RequestBody SignupRequest request) {
        log.info("Signup request for: {}", request.email());
        SignupResponse response = signupService.signup(request);
        return ResponseEntity.status(201).body(response);
    }

    /** 인증 실행 (Login) */
    @PostMapping("/authenticate")
    public ResponseEntity<AuthResponse> authenticate(
            @Valid @RequestBody AuthenticateRequest requestDto,
            HttpServletRequest request,
            HttpServletResponse response) {
        log.info("Authentication request for identifier: {}, type: {}", requestDto.identifier(),
                requestDto.credentialType());

        Credential providedCredential = credentialFactory.createFromRequest(requestDto);
        AuthenticationContext context = createAuthContext(request);

        AuthenticationApplicationService.AuthenticationResult result = authenticationService.authenticate(
                requestDto.identifier(),
                CredentialType.valueOf(requestDto.credentialType().toUpperCase()),
                providedCredential,
                context);

        if (result.authResult().authenticated()) {
            if (result.token() != null && result.token().refreshToken() != null) {
                setRefreshTokenCookie(response, result.token().refreshToken());
            }
        }

        return ResponseEntity.ok(AuthResponse.from(result.authResult(), result.token()));
    }

    /** MFA 완료 */
    @PostMapping("/mfa/complete")
    public ResponseEntity<AuthResponse> completeMfa(
            @Valid @RequestBody CompleteMfaRequest requestDto,
            HttpServletRequest request,
            HttpServletResponse response) {
        log.info("MFA completion request for session: {}", requestDto.mfaSessionId());

        MfaApplicationService.MfaCompletionResult result = mfaService.completeMfa(
                requestDto.mfaToken(),
                requestDto.mfaSessionId(),
                requestDto.mfaMethod(),
                requestDto.otpCode(),
                createAuthContext(request).getDevice());

        if (result.authResult().authenticated()) {
            if (result.token() != null && result.token().refreshToken() != null) {
                setRefreshTokenCookie(response, result.token().refreshToken());
            }
        }

        return ResponseEntity.ok(AuthResponse.from(result.authResult(), result.token()));
    }

    /** 토큰 갱신 (Refresh Token Rotation) */
    @PostMapping("/refresh")
    public ResponseEntity<AuthResponse> refresh(
            @CookieValue(value = REFRESH_TOKEN_COOKIE, required = false) String refreshToken,
            HttpServletRequest request,
            HttpServletResponse response) {
        log.info("Token refresh request");

        if (refreshToken == null || refreshToken.isEmpty()) {
            return ResponseEntity.status(401).build();
        }

        AuthenticationApplicationService.AuthenticationResult result =
                authenticationService.refreshToken(refreshToken, createAuthContext(request).getDevice());

        if (result.authResult().authenticated()) {
            if (result.token() != null && result.token().refreshToken() != null) {
                setRefreshTokenCookie(response, result.token().refreshToken());
            }
        } else {
            return ResponseEntity.status(401).body(AuthResponse.from(result.authResult(), null));
        }

        return ResponseEntity.ok(AuthResponse.from(result.authResult(), result.token()));
    }

    /** 로그아웃 */
    @PostMapping("/logout")
    public ResponseEntity<Void> logout(
            @CookieValue(value = REFRESH_TOKEN_COOKIE, required = false) String refreshToken,
            HttpServletResponse response) {
        log.info("Logout request");

        authenticationService.logout(refreshToken);
        deleteRefreshTokenCookie(response);

        return ResponseEntity.noContent().build();
    }

    private AuthenticationContext createAuthContext(HttpServletRequest request) {
        String ipAddress = request.getRemoteAddr();
        String userAgent = request.getHeader("User-Agent");
        String channel = request.getHeader("X-Channel");
        if (channel == null) channel = "WEB";

        return new AuthenticationContext(
                ipAddress,
                userAgent != null ? userAgent : "Unknown",
                channel,
                Instant.now(),
                false);
    }

    private void setRefreshTokenCookie(HttpServletResponse response, String refreshToken) {
        ResponseCookie cookie = ResponseCookie.from(REFRESH_TOKEN_COOKIE, refreshToken)
                .httpOnly(true)
                .secure(true) // Production에서는 true
                .path("/")
                .maxAge(7 * 24 * 60 * 60) // 7일
                .sameSite("Strict")
                .build();
        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());
    }

    private void deleteRefreshTokenCookie(HttpServletResponse response) {
        ResponseCookie cookie = ResponseCookie.from(REFRESH_TOKEN_COOKIE, "")
                .httpOnly(true)
                .path("/")
                .maxAge(0)
                .build();
        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());
    }
}

File: apps/auth-server/src/main/kotlin/com/example/chat/auth/server/api/AuthApi.kt
package com.example.chat.auth.server.api

import com.example.chat.auth.server.api.dto.factory.CredentialFactory
import com.example.chat.auth.server.api.dto.request.AuthenticateRequest
import com.example.chat.auth.server.api.dto.request.CompleteMfaRequest
import com.example.chat.auth.server.api.dto.request.SignupRequest
import com.example.chat.auth.server.api.dto.response.AuthResponse
import com.example.chat.auth.server.api.dto.response.SignupResponse
import com.example.chat.auth.server.application.service.AuthenticationApplicationService
import com.example.chat.auth.server.application.service.MfaApplicationService
import com.example.chat.auth.server.application.service.SignupApplicationService
import com.example.chat.auth.server.core.domain.AuthenticationContext
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import jakarta.validation.Valid
import org.springframework.http.HttpHeaders
import org.springframework.http.ResponseCookie
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.time.Instant

private val log = KotlinLogging.logger {}
private const val REFRESH_TOKEN_COOKIE = "refresh_token"

@RestController
@RequestMapping("/api/v1/auth")
class AuthApi(
        private val authenticationService: AuthenticationApplicationService,
        private val mfaService: MfaApplicationService,
        private val credentialFactory: CredentialFactory,
        private val signupService: SignupApplicationService
) {

    /** 회원가입 */
    @PostMapping("/signup")
    fun signup(
            @Valid @RequestBody request: SignupRequest
    ): ResponseEntity<SignupResponse> {
        log.info { "Signup request for: ${request.email}" }
        val response = signupService.signup(request)
        return ResponseEntity.status(201).body(response)
    }

    /** 인증 실행 (Login) */
    @PostMapping("/authenticate")
    fun authenticate(
            @Valid @RequestBody requestDto: AuthenticateRequest,
            request: HttpServletRequest,
            response: HttpServletResponse
    ): ResponseEntity<AuthResponse> {

        log.info { "${"Authentication request for identifier: {}, type: {}"} ${requestDto.identifier} ${requestDto.credentialType}" }

        val providedCredential = credentialFactory.createFromRequest(requestDto)
        val context = createAuthContext(request)

        val result =
                authenticationService.authenticate(
                        requestDto.identifier!!,
                        requestDto.credentialType!!,
                        providedCredential,
                        context
                )

        if (result.authResult.isAuthenticated) {
            result.token?.refreshToken?.let { setRefreshTokenCookie(response, it) }
        }

        return ResponseEntity.ok(AuthResponse.from(result.authResult, result.token))
    }

    /** MFA 완료 */
    @PostMapping("/mfa/complete")
    fun completeMfa(
            @Valid @RequestBody requestDto: CompleteMfaRequest,
            request: HttpServletRequest,
            response: HttpServletResponse
    ): ResponseEntity<AuthResponse> {

        log.info { "${"MFA completion request for session: {}"} ${requestDto.mfaSessionId}" }

        val result =
                mfaService.completeMfa(
                        requestDto.mfaToken!!,
                        requestDto.mfaSessionId!!,
                        requestDto.mfaMethod!!,
                        requestDto.otpCode!!,
                        createAuthContext(request).getDevice()
                )

        if (result.authResult.isAuthenticated) {
            result.token?.refreshToken?.let { setRefreshTokenCookie(response, it) }
        }

        return ResponseEntity.ok(AuthResponse.from(result.authResult, result.token))
    }

    /** 토큰 갱신 (Refresh Token Rotation) */
    @PostMapping("/refresh")
    fun refresh(
            @CookieValue(value = REFRESH_TOKEN_COOKIE, required = false) refreshToken: String?,
            request: HttpServletRequest,
            response: HttpServletResponse
    ): ResponseEntity<AuthResponse> {

        log.info { "Token refresh request" }

        if (refreshToken.isNullOrEmpty()) {
            return ResponseEntity.status(401).build()
        }

        val result =
                authenticationService.refreshToken(
                    refreshTokenVal = refreshToken,
                    device = createAuthContext(request).getDevice()
                )

        if (result.authResult.isAuthenticated) {
            result.token?.refreshToken?.let { setRefreshTokenCookie(response, it) }
        } else {
            return ResponseEntity.status(401).body(AuthResponse.from(result.authResult, null))
        }

        return ResponseEntity.ok(AuthResponse.from(result.authResult, result.token))
    }

    /** 로그아웃 */
    @PostMapping("/logout")
    fun logout(
            @CookieValue(value = REFRESH_TOKEN_COOKIE, required = false) refreshToken: String?,
            response: HttpServletResponse
    ): ResponseEntity<Void> {

        log.info { "Logout request" }

        authenticationService.logout(refreshToken)
        deleteRefreshTokenCookie(response)

        return ResponseEntity.noContent().build()
    }

    private fun createAuthContext(request: HttpServletRequest): AuthenticationContext {
        val ipAddress = request.remoteAddr
        val userAgent = request.getHeader("User-Agent")
        var channel = request.getHeader("X-Channel")
        if (channel == null) channel = "WEB"

        return AuthenticationContext(
                ipAddress = ipAddress,
                userAgent = userAgent ?: "Unknown",
                channel = channel,
                attemptTime = Instant.now(),
                isSuspiciousActivity = false
        )
    }

    private fun setRefreshTokenCookie(response: HttpServletResponse, refreshToken: String) {
        val cookie =
                ResponseCookie.from(REFRESH_TOKEN_COOKIE, refreshToken)
                        .httpOnly(true)
                        .secure(true) // Production에서는 true
                        .path("/")
                        .maxAge(7 * 24 * 60 * 60) // 7일
                        .sameSite("Strict")
                        .build()
        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString())
    }

    private fun deleteRefreshTokenCookie(response: HttpServletResponse) {
        val cookie =
                ResponseCookie.from(REFRESH_TOKEN_COOKIE, "")
                        .httpOnly(true)
                        .path("/")
                        .maxAge(0)
                        .build()
        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString())
    }
}
